quipucords/api/aggregate_report/model.py

```diff
@@ -0,0 +1,275 @@
+            aggregated.jboss_ws_cores_physical += jboss_ws_cpu_core_count
+        elif fingerprint.infrastructure_type == SystemFingerprint.HYPERVISOR:
+            aggregated.instances_hypervisor += 1
+        else:
+            aggregated.instances_unknown += 1
+        aggregated.jboss_eap_instances += 1 if jboss_eap_present else 0
+        aggregated.jboss_ws_instances += 1 if jboss_ws_present else 0
+
+        if fingerprint.cpu_socket_count:
+            # This may look strange, but I have been assured that it is correct.
+            # Some products are sold specifically with this per-two-CPU-sockets math.
+            # Try not to think too hard about this.
+            aggregated.socket_pairs += ceil(fingerprint.cpu_socket_count / 2)
+
+        if source_type == DataSources.NETWORK:
+            if not fingerprint.system_purpose:
+                aggregated.missing_system_purpose += 1
+            if not fingerprint.redhat_certs:
+                aggregated.missing_pem_files += 1
+        elif source_type == DataSources.OPENSHIFT:
+            if fingerprint.cpu_count:
+                # cpu_count originally comes from raw fact ["node"]["capacity"]["cpu"].
+                # A similar metric exists in raw fact ["node_metrics"]["package"].
+                # TODO Which of these values is more trustworthy?
+                aggregated.openshift_cores += fingerprint.cpu_count
+        elif source_type == DataSources.VCENTER:
+            vmware_cluster = fingerprint.vm_cluster
+            vmware_host = fingerprint.virtual_host_uuid
+            if vmware_cluster and vmware_host:
+                vmware_vms_by_host[vmware_cluster].append(vmware_host)
+
+    for os_name, os_versions in os_by_name_and_version.items():
+        aggregated.os_by_name_and_version[os_name] = dict(os_versions)
+
+    aggregated.system_creation_date_average = average_date(system_creation_dates)
+
+    if vmware_vms_by_host:
+        # By collecting *all* the hosts in vmware_vms_by_host, we can effectively
+        # deduplicate them here. We collected the hosts and VMs by their names,
+        # not their UUIDs, which means that if two actually different hosts or VMs
+        # have the same name (e.g. "localhost"), then we might under-count them here.
+        # Justin said this is correct and consistent with how other aspects of our
+        # business count usage by simple hostnames.
+        vmware_vm_count = sum(len(set(items)) for items in vmware_vms_by_host.values())
+        vmware_host_count = len(vmware_vms_by_host)
+        aggregated.vmware_vms = vmware_vm_count
+        aggregated.vmware_vm_to_host_ratio = vmware_vm_count / vmware_host_count
+        aggregated.vmware_hosts = vmware_host_count
+
+    #####
+    # Aggregate more data from raw facts, not fingerprints. Hic sunt dracones!
+    # Can any of these actually be determined by fields in SystemFingerprint?
+    # Should we normalize and store some of these as fields in SystemFingerprint?
```

Comment:
We could. The 2yrs old problem here is: data from those sources won't fit SystemFingerprint. I say we make this part of the "shiny new normalization", and as part of the process we create the proper models to hold their data.

Openshift deserves normalization. Ansible and RHACS I'm not so sure.

I'd even say: remove normalization as a requirement for discovery. If a certain source has no need for it, don't do it. Of course we could also create specialized models just for the normalized result of those sources.